File: apps/linguist/src/astra_linguist/campaigns.py
"""Campaign matching + billing + context (P4, port of `campaigns.ts`).

The matching/billing/context LOGIC is linguist's; the campaign DATA is
ontology-being (G3). ontology's `Role{player(slug), character, descriptions}` is
adapted into faerrin's `roles[playerDisplayName] -> CharacterRole[]` view (the
billing/context text bills players by display name, so the slug→name map matters).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from astra_ontology.models import Being

from .models import Transcript

GM_NAMES = frozenset({"Gamemaster", "Dungeon Master"})
# faerrin config.ts `campaign` — kept here as linguist tuning constants.
MATCH_THRESHOLD = 15
TRANSCRIPTION_CONFIDENCE = "85%"


@dataclass(frozen=True)
class CharacterRole:
    """A character a player can be billed as: display name + description facts."""

    name: str
    desc: list[str]


@dataclass(frozen=True)
class CampaignView:
    """faerrin-shaped campaign: roles keyed by player **display name**."""

    name: str
    is_main: bool
    roles: dict[str, list[CharacterRole]]


@dataclass(frozen=True)
class MatchedCampaign:
    campaign: CampaignView
    idx: int  # position in the list → filename prefix
    billing: dict[str, CharacterRole]  # player display name → billed character


def campaign_views(being: Being) -> list[CampaignView]:
    """Adapt ontology-being campaigns into the faerrin role-keyed view."""
    slug_to_name = {player.slug: player.name for player in being.players}
    views: list[CampaignView] = []
    for campaign in being.campaigns:
        roles: dict[str, list[CharacterRole]] = {}
        for role in campaign.roles:
            player = slug_to_name.get(role.player, role.player)
            roles.setdefault(player, []).append(
                CharacterRole(name=role.character, desc=list(role.descriptions))
            )
        views.append(CampaignView(name=campaign.name, is_main=campaign.main, roles=roles))
    return views


def _character_names(campaign: CampaignView) -> list[str]:
    """All non-GM character names — the keyword set for matching."""
    return [
        role.name
        for roles in campaign.roles.values()
        for role in roles
        if role.name not in GM_NAMES
    ]


def match_campaign(
    transcript: Transcript,
    campaigns: list[CampaignView],
    *,
    threshold: int = MATCH_THRESHOLD,
) -> MatchedCampaign | None:
    """First campaign whose character names clear the keyword threshold + its billing."""
    for idx, campaign in enumerate(campaigns):
        keywords = _character_names(campaign)
        hits: dict[str, int] = {kw: 0 for kw in keywords}
        for line in transcript.script:
            for keyword in keywords:
                if keyword in line.text:
                    hits[keyword] += 1

        if sum(hits.values()) < threshold:
            continue

        billing: dict[str, CharacterRole] = {}
        for player, roles in campaign.roles.items():
            best_name: str | None = None
            best = -1
            for role in roles:
                hit = hits.get(role.name, -1)
                if hit > best:
                    best_name, best = role.name, hit
            for role in roles:
                if role.name == best_name or role.name in GM_NAMES:
                    billing[player] = role
                    break
        return MatchedCampaign(campaign=campaign, idx=idx, billing=billing)
    return None


def campaign_filename(matched: MatchedCampaign) -> str:
    """Output stem, e.g. `000.through-a-song-darkly`."""
    ident = "0" if matched.campaign.is_main else "1"
    idx = f"{matched.idx:02d}"
    name = matched.campaign.name.lower().replace(" ", "-").replace(",", "")
    return f"{ident}{idx}.{name}"


def _render_desc(desc: list[str]) -> str:
    """Single fact inline; multiple facts as an indented dash list (faerrin shibboleth)."""
    if len(desc) == 1:
        return desc[0]
    return "".join(f"\n    - {fact}" for fact in desc)


def make_billing(matched: MatchedCampaign) -> str:
    billings = [
        f"The role of {role.name} is played by the player {player}."
        for player, role in matched.billing.items()
    ]
    return "\n".join(["Billing:\n", *billings])


def make_context(matched: MatchedCampaign, date: str) -> str:
    session_kind = (
        f'This is from the main campaign of the game, "{matched.campaign.name}".'
        if matched.campaign.is_main
        else f'This is from a one-shot side story of the game, "{matched.campaign.name}".'
    )
    descs = [
        f"  - {role.name}: {_render_desc(role.desc)}"
        for role in matched.billing.values()
        if role.name not in GM_NAMES
    ]
    # The fixed instruction paragraph — byte-identical to faerrin's template
    # (split across source lines only; the value, incl. internal `\n`, is unchanged).
    paragraph = (
        "The first portion of the transcript is a pre-session chat between the players,"
        " and then it transitions to the\nactual session around when someone says something"
        ' similar to "Do we want to play Pathfinder?" or "Does someone\nwant to do a recap?".'
        ' The session ends when the Gamemaster says somthing similar to "Do we want to call'
        ' it there\nfor the evening?" near the end of the transcript.'
    )
    return (
        "Context:\n\n"
        "This is a transcript of an ongoing ttrpg game in the setting of Faerrin, "
        f"recorded on {date}. The four player characters are:\n"
        f"{chr(10).join(descs)}\n\n"
        f"{session_kind}\n\n"
        f"{paragraph}\n\n"
        "Note: transcription is machine-recorded, with an "
        f"{TRANSCRIPTION_CONFIDENCE} confidence rate.\n"
    )


def to_shibboleth(campaigns: list[CampaignView]) -> dict[str, dict[str, Any]]:
    """Serialize to the shibboleth.json shape (object keyed by campaign name)."""
    return {
        campaign.name: {
            "isMain": campaign.is_main,
            "roles": {
                player: [{"name": role.name, "desc": role.desc} for role in roles]
                for player, roles in campaign.roles.items()
            },
        }
        for campaign in campaigns
    }

File: apps/akasha-frontend/src/domain/lib/campaigns.ts
/**
 * Campaign matcher — ported from faerrin `pkg/content/scripts/lib/campaigns.ts`
 * (the content heuristic that routes each session into its `Script/<campaign>/`
 * folder), adapted to astra's `@astra/ontology` Campaign shape: roles are a flat
 * `Role[]` of `{player, character, character_class}` (vs faerrin's
 * `roles: Record<player, CharacterRole[]>`), and `role.player` is the player SLUG
 * (e.g. "jorge") so billing is re-keyed to the display NAME (e.g. "Jorge") to match
 * the transcript's `user.name`.
 *
 * Behaviour is byte-faithful: count non-GM character-name string hits across the
 * transcript; the FIRST campaign — in being.kdl order — whose total clears
 * `MATCH_THRESHOLD` wins; else null → Unsorted. This is URL-parity-critical (N7):
 * a wrong match misfiles the session's page. Guarded by the slug-parity fixture.
 */
import type { Campaign, Role } from "@astra/ontology";
import type { Transcript } from "./transcript";

// faerrin `pkg/content/scripts/config.ts` campaign.matchThreshold. A content-heuristic
// tuning constant (not deploy config), so it lives in code like slug.ts's regexes.
const MATCH_THRESHOLD = 15;

/** GM roles are excluded from the keyword set (faerrin excluded "Gamemaster"/
 *  "Dungeon Master" by name; astra marks them character_class="gm"). */
const isGm = (r: Role): boolean => r.character_class === "gm";

export interface MatchedCampaign {
  campaign: Campaign;
  /** display name → the character that player is billed as this session (name toggle). */
  billing: Record<string, string>;
}

function characterNames(campaign: Campaign): string[] {
  return campaign.roles.filter((r) => !isGm(r)).map((r) => r.character);
}

/**
 * @param nameBySlug player slug → display name (from being.players), so `billing`
 *   is keyed by the same name the transcript's `user.name` carries.
 */
export function matchCampaign(
  transcript: Transcript,
  campaigns: Campaign[],
  nameBySlug: Map<string, string>,
): MatchedCampaign | null {
  for (const campaign of campaigns) {
    const keywords = characterNames(campaign);
    const hits: Record<string, number> = {};
    for (const k of keywords) hits[k] = 0;
    for (const { text } of transcript.script) {
      for (const keyword of keywords) {
        if (text.includes(keyword)) hits[keyword] = (hits[keyword] ?? 0) + 1;
      }
    }

    let sum = 0;
    for (const v of Object.values(hits)) sum += v;
    if (sum < MATCH_THRESHOLD) continue;

    // Bill each player as their highest-hit character (GM stays GM), keyed by name.
    const byPlayer = new Map<string, Role[]>();
    for (const r of campaign.roles) {
      const list = byPlayer.get(r.player) ?? [];
      list.push(r);
      byPlayer.set(r.player, list);
    }
    const billing: Record<string, string> = {};
    for (const [playerSlug, roles] of byPlayer) {
      const name = nameBySlug.get(playerSlug) ?? playerSlug;
      const gm = roles.find(isGm);
      if (gm) {
        billing[name] = gm.character;
        continue;
      }
      let best = -1;
      let bestChar = roles[0]?.character ?? name;
      for (const r of roles) {
        const h = hits[r.character] ?? -1;
        if (h > best) {
          best = h;
          bestChar = r.character;
        }
      }
      billing[name] = bestChar;
    }
    return { campaign, billing };
  }
  return null;
}

/** The character a speaker is billed as this session, or their real name if the
 *  session matched no campaign (or they aren't a billed player). */
export function characterFor(realName: string, match: MatchedCampaign | null): string {
  return match?.billing[realName] ?? realName;
}
